Add Google Cloud Storage support

mvanholsteijn opened the pull request:
This PR adds Google Cloud Storage backend support to privatebin, allowing for a server less deployment.

## Changes

* added a GoogleCloudStorage Data class
* added unit tests for GoogleCloudStorage
* added the Google Storage Client API library
* added the configuration options for the GoogleCloudStorage


Closes #794

elrido commented:
Thank you for the effort, this is looking pretty good on first glance. I'll now take a closer look at the new class to provide further feedback, if necessary.

In the mean time, could you re-run composer as follows in your branch, please? If easy to do, please remove the commit that added them to the history, but I can also handle that via a `git merge --squash` upon merging the branch.
```
composer update --no-dev --optimize-autoloader
```
We don't commit the development dependencies to the repository, as they are useless when deployed in production and increase the size of the repository by a massive amount. Part of our user base use the git repository as their deployment mechanism.

elrido commented on .gitignore:
```diff
@@ -7,6 +7,8 @@ cfg/*
 
 # Ignore data/
 data/
+!lib/Data
```
I assume these had to be added due to an OS using case-insensitive file system? Would it maybe be cleaner to rewrite the rule before this one to an absolute path (`/data/` instead of `data/`), so it doesn't match the {lib,tst}/Data folders?

mvanholsteijn replied:
done. I changed it to /data/.

elrido commented on composer.json:
```diff
@@ -27,7 +27,8 @@
 		"yzalis/identicon" : "2.0.0",
-		"mlocati/ip-lib" : "1.14.0"
+		"mlocati/ip-lib" : "1.14.0",
+		"google/cloud-storage": "^1.23"
```
Very minor: We use explicit, precide versions to avoid accidental upgrades that are overlooked when working on other dependencies and therefore not documented in the changelog. And it is always 100% clear what version got used at any point in time.

mvanholsteijn replied:
done

elrido commented on lib/Data/GoogleCloudStorage.php:
```diff
@@ -0,0 +1,224 @@
+            $metadata = array_key_exists('meta', $comment) ? $comment['meta'] : [];
+            unset($metadata['attachment'], $metadata['attachmentname'], $metadata['salt']);
+
+            $this->_bucket->upload(json_encode($comment), [
```
Would it be possible to use `Json::encode()` instead of `json_encode()`? It is a very thin wrapper that detects and throws an exception when an error gets detected, by calling `json_last_error()` to detect errors of the `json_encode()` function, which otherwise might silently return a partial or even empty string.

elrido commented on lib/Data/GoogleCloudStorage.php:
```diff
@@ -0,0 +1,224 @@
+                    'metadata' => $metadata,
+                ]
+            ]);
+        } catch (\InvalidArgumentException $e) {
```
Please add a:
```
use InvalidArgumentException;
```
and remove the `\` here.

This is of course purely stylistic, but follows one of the PSR guidelines we try to stick to, for consistency in our coding style.

mvanholsteijn replied:
done

elrido reviewed (changes requested):
Apart from the comments made on specific lines, please [apply the patch file (found via the "raw diff" button) from StyleCI](https://github.styleci.io/analyses/bQ7vna) to apply our coding guidelines.

Furthermore, it seems the unit tests fail in Scrutinizer or Github actions as well as for me locally. The error message is `Google\Cloud\Core\Exception\GoogleException: No project ID was provided, and we were unable to detect a default project ID. in /home/scrutinizer/build/vendor/google/cloud-storage/src/StorageClient.php:571`. Is it possible to turn on a a mock mode or use mock classes in the Google cloud storage library, so that it emulates the API operations without actually accessing anything over the network? For the database storage we use an in-memory sqlite database, for example.

elrido commented:
Hope this helps - I found some examples of mocking local files to appear being stored on gcs in their unit tests:
https://github.com/googleapis/google-cloud-php-storage/blob/ca3817dae90d7d757381f3f63c9d7df58ca1416f/tests/Unit/StreamWrapperTest.php#L453-L461

mvanholsteijn commented:
> Hope this helps - I found some examples of mocking local files to appear being stored on gcs in their unit tests:
> https://github.com/googleapis/google-cloud-php-storage/blob/ca3817dae90d7d757381f3f63c9d7df58ca1416f/tests/Unit/StreamWrapperTest.php#L453-L461

Mocking all the GCS api calls, does not feel like it is going to delivery much value. The mocks will just be white box tests for the implementation I wrote. When I copied the unit test of filesystem to point to GCS, it found quite a few errors.

Unfortunately there is no in memory version at the moment.

mvanholsteijn commented:
Both Scrutinizer and github actions do support starting a service container for the build/test process. Where can I edit the .scrutinize.yaml ?

https://scrutinizer-ci.com/docs/build/services

rugk commented:
Approved your PR now, so the stuff can run.

elrido commented:
For the local unit testing, I use our dedicated [container image](https://github.com/PrivateBin/docker-unit-testing) and will see if I get the fake-gcs-server binary integrated into it, so it gets started up in the background on phpunit runs and torn down again. Then I could further test the branch locally and use the coverage reports to understand the logic.

If it isn't easy to get it to work in scrutinizer, we can always resort to skipping this one test class for it's test-runs. To me, the local testing ability is much more important, as I rely on it for my test-change-test-commit cycles when working on the project.

elrido commented:
> Where can I edit the .scrutinize.yaml ?

Not sure why this was only in their web backend configuration - maybe they didn't yet support the in-repo-config-file when we first set this up? In any case, I've exported our current settings and added them to the master branch at https://github.com/PrivateBin/PrivateBin/blob/dae093bbfa73fff54ae6521d81d6d5f4bbd0d226/.scrutinizer.yml

elrido commented:
I've added the fake service into the unit test image in https://github.com/PrivateBin/docker-unit-testing/blob/master/unit-test.sh#L20 and running the unit tests locally it gets further, but still fails as follows:
```
1) GoogleCloudStorageTest
InvalidArgumentException: json_decode error: Syntax error in /tmp/repo/vendor/google/cloud-core/src/JsonTrait.php:40
Stack trace:
#0 /tmp/repo/vendor/google/cloud-core/src/RequestBuilder.php(161): Google\Cloud\Core\RequestBuilder::jsonDecode()
#1 /tmp/repo/vendor/google/cloud-core/src/RequestBuilder.php(60): Google\Cloud\Core\RequestBuilder->loadServiceDefinition()
#2 /tmp/repo/vendor/google/cloud-storage/src/Connection/Rest.php(93): Google\Cloud\Core\RequestBuilder->__construct()
#3 /tmp/repo/vendor/google/cloud-storage/src/StorageClient.php(106): Google\Cloud\Storage\Connection\Rest->__construct()
#4 /tmp/repo/tst/Data/GoogleCloudStorageTest.php(26): Google\Cloud\Storage\StorageClient->__construct()
#5 /usr/local/vendor/phpunit/phpunit/src/Framework/TestSuite.php(679): GoogleCloudStorageTest::setUpBeforeClass()
#6 /usr/local/vendor/phpunit/phpunit/src/Framework/TestSuite.php(733): PHPUnit_Framework_TestSuite->run()
#7 /usr/local/vendor/phpunit/phpunit/src/TextUI/TestRunner.php(517): PHPUnit_Framework_TestSuite->run()
#8 /usr/local/vendor/phpunit/phpunit/src/TextUI/Command.php(186): PHPUnit_TextUI_TestRunner->doRun()
#9 /usr/local/vendor/phpunit/phpunit/src/TextUI/Command.php(116): PHPUnit_TextUI_Command->run()
#10 /usr/local/vendor/phpunit/phpunit/phpunit(52): PHPUnit_TextUI_Command::main()
#11 {main}
2) GoogleCloudStorageTest
InvalidArgumentException: json_decode error: Syntax error in /tmp/repo/vendor/google/cloud-core/src/JsonTrait.php:40
Stack trace:
#0 /tmp/repo/vendor/google/cloud-core/src/RequestBuilder.php(161): Google\Cloud\Core\RequestBuilder::jsonDecode()
#1 /tmp/repo/vendor/google/cloud-core/src/RequestBuilder.php(60): Google\Cloud\Core\RequestBuilder->loadServiceDefinition()
#2 /tmp/repo/vendor/google/cloud-storage/src/Connection/Rest.php(93): Google\Cloud\Core\RequestBuilder->__construct()
#3 /tmp/repo/vendor/google/cloud-storage/src/StorageClient.php(106): Google\Cloud\Storage\Connection\Rest->__construct()
#4 /tmp/repo/tst/Data/GoogleCloudStorageTest.php(26): Google\Cloud\Storage\StorageClient->__construct()
#5 /usr/local/vendor/phpunit/phpunit/src/Framework/TestSuite.php(679): GoogleCloudStorageTest::setUpBeforeClass()
#6 /usr/local/vendor/phpunit/phpunit/src/Framework/TestSuite.php(733): PHPUnit_Framework_TestSuite->run()
#7 /usr/local/vendor/phpunit/phpunit/src/TextUI/TestRunner.php(517): PHPUnit_Framework_TestSuite->run()
#8 /usr/local/vendor/phpunit/phpunit/src/TextUI/Command.php(186): PHPUnit_TextUI_TestRunner->doRun()
#9 /usr/local/vendor/phpunit/phpunit/src/TextUI/Command.php(116): PHPUnit_TextUI_Command->run()
#10 /usr/local/vendor/phpunit/phpunit/phpunit(52): PHPUnit_TextUI_Command::main()
#11 {main}
3) GoogleCloudStorageTest
InvalidArgumentException: json_decode error: Syntax error in /tmp/repo/vendor/google/cloud-core/src/JsonTrait.php:40
Stack trace:
#0 /tmp/repo/vendor/google/cloud-core/src/RequestBuilder.php(161): Google\Cloud\Core\RequestBuilder::jsonDecode()
#1 /tmp/repo/vendor/google/cloud-core/src/RequestBuilder.php(60): Google\Cloud\Core\RequestBuilder->loadServiceDefinition()
#2 /tmp/repo/vendor/google/cloud-storage/src/Connection/Rest.php(93): Google\Cloud\Core\RequestBuilder->__construct()
#3 /tmp/repo/vendor/google/cloud-storage/src/StorageClient.php(106): Google\Cloud\Storage\Connection\Rest->__construct()
#4 /tmp/repo/tst/Data/GoogleCloudStorageTest.php(26): Google\Cloud\Storage\StorageClient->__construct()
#5 /usr/local/vendor/phpunit/phpunit/src/Framework/TestSuite.php(679): GoogleCloudStorageTest::setUpBeforeClass()
#6 /usr/local/vendor/phpunit/phpunit/src/Framework/TestSuite.php(733): PHPUnit_Framework_TestSuite->run()
#7 /usr/local/vendor/phpunit/phpunit/src/TextUI/TestRunner.php(517): PHPUnit_Framework_TestSuite->run()
#8 /usr/local/vendor/phpunit/phpunit/src/TextUI/Command.php(186): PHPUnit_TextUI_TestRunner->doRun()
#9 /usr/local/vendor/phpunit/phpunit/src/TextUI/Command.php(116): PHPUnit_TextUI_Command->run()
#10 /usr/local/vendor/phpunit/phpunit/phpunit(52): PHPUnit_TextUI_Command::main()
#11 {main}
4) GoogleCloudStorageTest
InvalidArgumentException: json_decode error: Syntax error in /tmp/repo/vendor/google/cloud-core/src/JsonTrait.php:40
Stack trace:
#0 /tmp/repo/vendor/google/cloud-core/src/RequestBuilder.php(161): Google\Cloud\Core\RequestBuilder::jsonDecode()
#1 /tmp/repo/vendor/google/cloud-core/src/RequestBuilder.php(60): Google\Cloud\Core\RequestBuilder->loadServiceDefinition()
#2 /tmp/repo/vendor/google/cloud-storage/src/Connection/Rest.php(93): Google\Cloud\Core\RequestBuilder->__construct()
#3 /tmp/repo/vendor/google/cloud-storage/src/StorageClient.php(106): Google\Cloud\Storage\Connection\Rest->__construct()
#4 /tmp/repo/tst/Data/GoogleCloudStorageTest.php(26): Google\Cloud\Storage\StorageClient->__construct()
#5 /usr/local/vendor/phpunit/phpunit/src/Framework/TestSuite.php(679): GoogleCloudStorageTest::setUpBeforeClass()
#6 /usr/local/vendor/phpunit/phpunit/src/Framework/TestSuite.php(733): PHPUnit_Framework_TestSuite->run()
#7 /usr/local/vendor/phpunit/phpunit/src/TextUI/TestRunner.php(517): PHPUnit_Framework_TestSuite->run()
#8 /usr/local/vendor/phpunit/phpunit/src/TextUI/Command.php(186): PHPUnit_TextUI_TestRunner->doRun()
#9 /usr/local/vendor/phpunit/phpunit/src/TextUI/Command.php(116): PHPUnit_TextUI_Command->run()
#10 /usr/local/vendor/phpunit/phpunit/phpunit(52): PHPUnit_TextUI_Command::main()
#11 {main}
```
I think the tests already point to the right place (localhost:4443). I tried running the fake service both in http and https mode. Any other parameters that would need to be passed to it?

mvanholsteijn commented:
https://github.com/PrivateBin/PrivateBin/pull/795#issuecomment-847011274

Superweird.  When I run inside the unit-test container, I get the error message:

```
2) GoogleCloudStorageTest
Google\Cloud\Core\Exception\ServiceException: Connection refused for URI https://localhost:4443/storage/v1/b?project=fake-project&prettyPrint=false in /tmp/repo/vendor/google/cloud-core/src/RequestWrapper.php:369
Stack trace:
#0 /tmp/repo/vendor/google/cloud-core/src/RequestWrapper.php(208): Google\Cloud\Core\RequestWrapper->convertToGoogleException()
#1 /tmp/repo/vendor/google/cloud-core/src/RestTrait.php(97): Google\Cloud\Core\RequestWrapper->send()
#2 /tmp/repo/vendor/google/cloud-storage/src/Connection/Rest.php(176): Google\Cloud\Storage\Connection\Rest->send()
#3 /tmp/repo/vendor/google/cloud-storage/src/StorageClient.php(347): Google\Cloud\Storage\Connection\Rest->insertBucket()
#4 /tmp/repo/tst/Data/GoogleCloudStorageTest.php(28): Google\Cloud\Storage\StorageClient->createBucket()
#5 /tmp/repo/vendor/phpunit/phpunit/src/Framework/TestSuite.php(679): GoogleCloudStorageTest::setUpBeforeClass()
#6 /tmp/repo/vendor/phpunit/phpunit/src/TextUI/TestRunner.php(517): PHPUnit_Framework_TestSuite->run()
#7 /usr/local/vendor/phpunit/phpunit/src/TextUI/Command.php(186): PHPUnit_TextUI_TestRunner->doRun()
#8 /usr/local/vendor/phpunit/phpunit/src/TextUI/Command.php(116): PHPUnit_TextUI_Command->run()
#9 /usr/local/vendor/phpunit/phpunit/phpunit(52): PHPUnit_TextUI_Command::main()
#10 {main}
```

When I run it locally on machine the result is good.

r4sas commented:
I have one question: are we really must take all vendor stuff (over 400 files) with project? Can't that requirement be added to composer as optional (using `suggest` section)?

elrido commented:
> are we really must take all vendor stuff (over 400 files) with project?

I do agree that adding 3 MiB of source to a project of 2 MiB is a very heavy change. I would wish there would be a more lightweight implementation for this API. The existing DB storage has less impact, in comparison, because the database access is handled in a php module, so we don't have to ship any code and it is up to the server admin to install/load the module, if they want to use it.

If we don't add these vendor files, it means that admins have to install and run composer to install it or that we have to implement tooling to our release process to generate archives that include these files. At the moment neither is necessary: git archive generates archives that contain everything that is necessary and some admins choose to use a git clone/pull to update their instance.

I'm open to any suggestions on how this could be addressed. Is there a GCS php module, for example, avoiding the need for a pure php implementation of this API?

mvanholsteijn commented:
- there is no PHP Module which implements the  GCS client library. 
- I am pretty happy to move 'google/cloud-storage' requirement to the "suggest" category; But I am not quite sure how this will impact your CI/CD processes.
- I know why the unit test fails: the library is trying to generate authentication tokens which are not available. I will think of a way to solve that.

rugk commented:
>But I am not quite sure how this will impact your CI/CD processes.

Well… CI processes can be adjusted. So this should not prevent you from having a clean repo/installation with only an optional dependency.

elrido commented:
Ah, I wasn't aware of the "suggest" option. So if it would get moved there, in the unit testing container we could provide these libraries pre-installed in the environment. For github actions and scrutinizer we could add/extend the composer step to also install the suggested packages (`composer update --install-suggests`). The installation instructions could mention the command in the section explaining GCS configuration.

mvanholsteijn commented:
@elrido 
1. The unit test completely independent of an external system by creating stubs for the used GCS api calls.
2. the google\cloud-storage dependency has moved to the suggest list.
Where can I fix the scrutinize configuration to install the library before analysis and unit tests start?

mvanholsteijn commented:
Gentlemen, I have made the google/cloud-storage package optional and the unit tests work within an external service.

How do I add the google/cloud-storage package to the distribution?

elrido commented:
>     2. the google\cloud-storage dependency has moved to the suggest list.
>        Where can I fix the scrutinize configuration to install the library before analysis and unit tests start?

> > [...] In any case, I've exported our current settings and added them to the master branch at https://github.com/PrivateBin/PrivateBin/blob/dae093bbfa73fff54ae6521d81d6d5f4bbd0d226/.scrutinizer.yml

> How do I add the google/cloud-storage package to the distribution?

If I understand the "suggest" option correctly, you would have to manually run the `composer require google/cloud-storage` in your installation folder. @r4sas, is that correct or is there a better way? I think the distribution (produced using `git archive`) won't include the composer files.

r4sas commented:
> @r4sas, is that correct or is there a better way?

Yes, it is.



> I think the distribution (produced using `git archive`) won't include the composer files.

In current `.gitattributres` - it will include them.

mvanholsteijn commented:
The use of the suggests list has quite a big impact and requires changes in the unit-test docker images, the ready-to-run docker image. the release process and the installation instructions.

In this PR I have removed the entire /vendor/ directory source control and added the google/cloud-storage library to the required list. I think it is the most simple solution, with the least impact, and provides a consist way of setting up a development environment.

I used the  [action-gh-release](https://github.com/softprops/action-gh-release) to create the distributable application and include the vendor libraries. The only thing missing is the GPG signature on the zip file.

Please let me know what you think.

elrido commented:
So, it is clear that either the release process or the unit testing needs to be changed for this feature.

IMHO the unit testing is very much under our control (local testing either via local phpunit / nodejs or docker container, in github actions and scrutinizer), it needs to be adapted anyway to add the API fake service and it affects only a hand full of people.

The release process on the other hand is not entirely under our control. Yes, the mainly advertised release is the archive github creates automatically on tags, but there are also folks that use the latest archive (which is git archive based) or git pulls. There are also a [handful of container images published on docker hub](https://hub.docker.com/search?q=privatebin&type=image), not maintained by us, that may need to change their build to include a composer step, depending on what they package.

The only thing at this point that I'd like to get working before I feel comfortable to merge this, is the unit testing in the container image. Should I not figure out how to get it working in scrutinizer or github actions, we can just rm that one test on those to skip it on there. As long as I can test it all locally I'm happy to get it merged and the image published for others to peruse.

Would it be possible to either revert the vendor folder removal or can you point me at the original branch before that commit got applied? In the mean time I can look into adding the one library into the unit testing image - I still have my older checkout of the branch to play with.

mvanholsteijn commented:
@elrido, I restored the /vendor/ directory for you and moved the google/cloud-storage library to the suggest list. 

- The unit test work by installing the library before running the tests.
- The GoogleCloudStorage sources are excluded from source code analysis as I cannot configure the suggest libraries to be installed before the source code is analysed.

I would appreciate it if you could help me with the instructions to add the google/cloud-storage library to the docker image `docker-nginx-fpm-alpine`.  When I do a `composer require google/cloud-storage --optimize-autoloader` in the /srv directory, it deletes all existing vendor files.

elrido commented:
Thank you, I'll look into latest by tomorrow.

Reg. the docker-nginx-fpm-alpine container image, I don't want to include this in the default image, but lets open a ticket over there and solve it via a tagged image. I recently added an edge tag and I think it's time to a) document all of the tags and b) split some of the functionalities. The default image includes the PDO modules, for example, and if we add a tag for gcs support, why not also add one without PDO modules for even lower size and attack surface - one tag per data storage type for each of pb release, nightly of pb master & pb release on alpine edge.

elrido commented:
As a follow up, I've raised a ticket for adding a container image flavour with GCS support baked in at [#41 in the image project](https://github.com/PrivateBin/docker-nginx-fpm-alpine/issues/41). I'd welcome any input on the naming scheme and while that discussion is ongoing, I can start the work on the different Dockerfiles and automation.
